content/docs/atom-guide/dev-guide/shaders/azsl/shader-variants-fallback-key.md

```diff
@@ -0,0 +1,134 @@
+    {
+    	float4 m_color; 
+    }
+    
+    PSOutput MainPS(VSOutput IN)
+    {
+        PSOutput OUT;
+        if (o_useObjectColor)
+        {
+            OUT.m_color = ObjectSrg::m_color;
+        }
+        else
+        {
+            OUT.m_color = DrawSrg::m_color;
+        }
+        return OUT;
+    }
+```
+And here is the solution, in which the developer chooses `DrawSrg` to be the Shader Variant Fallback. In other words, `DrawSrg` will own the Fallback Key:  
+```cpp
+    option bool o_useObjectColor;
+    
+    ShaderResourceGroupSemantic SRG_PerDraw
+    {
+        FrequencyId = 0;
+        ShaderVariantFallback = 128; // The SRG that gets this semantic will own the 128bits long Fallback Key.
+    }
+    ShaderResourceGroupSemantic SRG_PerObject
+    {
+        FrequencyId = 1;
+    }
+     
+    ShaderResourceGroup ObjectSrg : SRG_PerObject
+    {
+        float4 m_color;
+    }
+     
+    ShaderResourceGroup DrawSrg : SRG_PerDraw
+    {
+    	float4 m_color; 
+    }
+    
+    PSOutput MainPS(VSOutput IN)
+    {
+        PSOutput OUT;
+        if (o_useObjectColor)
+        {
+    	   OUT.m_color = ObjectSrg::m_color;
+        }
+        else
+        {
+           OUT.m_color = DrawSrg::m_color;
+        }
+        return OUT;
+    }
+```
+### How many bits long is this array of bits where the compiler encodes all of the Shader variant options?
+When one or more option variables are declared, one and only one SRG must have a fallback value.  
+  
+The size of the fallback value should be at least 128 bit.  
+The fallback value is specified on a Shader Resource Group Semantic.  
+```cpp
+    ShaderResourceGroupSemantic OptionExample
+    {
+        FrequencyId = 6;
+        ShaderVariantFallback = 128;
+    };
+```
+The Fallback Key is a Shader Constant, but has its own API to configure At runtime. For a 128 bits long fallback key, an array of four 32-bit integers will be reserved to hold the bitset representing the values of all options not defined statically at compilation time.  
+  
+Now that the concept of Fallback Key has been introduced, it is time to talk about Shader Variant Byte Code (*ShaderVariantAsset*) Permutations.  
+  
+## Permutations Of Shader Variant Assets.
+When a shader is compiled as an asset, at a minimum one *ShaderVariantAsset* is generated per platform (One for DX12, one for Vulkan, etc). This is called the Root Variant.  
+The Root Variant will be capable of executing any branch in the code at runtime because each Shader Variant Option will be used by the byte code.  
+Because the size of the Shader Variant Fallback Key is 128 bits, the Shader Developer can make use of up to 128 unique boolean options.  
+In order to fully "specialize" (or bake) all the possible permutations the shader build pipeline will have to generate (2<sup>128</sup>  - 1) ShaderVariantAssets, per platform.  
+Fortunately, the Shader Build Pipeline only builds the permutations specified by the developer.  
```

Comment:
```suggestion
When AZSLc compiles shader code and produces a shader asset, a *root variant* is generated per platform (DX12, Vulkan, Metal, and so on). The root variant can execute any branch in the code at runtime because the byte coded uses each shader variant option. Because the size of the shader variant fallback key is 128 bits, the shader developer can use up to 128 unique boolean options.

To fully specialize (or bake) all the possible permutations, the shader build pipeline must generate (2<sup>128</sup>  - 1) shader variant assets per platform. You may not need all those shader variant assets, so the the developer can specify which permutations the shader build pipeline builds. 

```

Reply:
Fixed.